src/Type/Generic/TemplateMixedType.php

```diff
@@ -123,7 +124,7 @@ public function inferTemplateTypes(Type $receivedType): TemplateTypeMap
 		}
 
 		return new TemplateTypeMap([
-			$this->name => $receivedType,
+			$this->name => TypeUtils::generalizeType($receivedType),
```

Comment:
Although this was not required when calling generic functions, generalizing the inferred types is needed for generic classes.

```
/**
 * @template T
 */
class C {
    /** @param T $a */
    public function __construct($a) {}

    /** @return T */
    public function get();

    /** @param T $a */
    public function set($a);
}
```

```
new C(1); // We want this to be C<int>, because we want the method signatures to be get(): int and set(int)
```